nni/contrib/distillation/preprocessor.py

```diff
@@ -0,0 +1,184 @@
+            And in some training methods, such as data augumentation, each time the same meta data selected from dataset
+            might be different after augumentation.
+            At this time, setting `>1` can run more epochs to obtain multiple epochs of distillation labels.
+        log_process
+            If logging during generate labels.
+        labels_split_fn
+            A function used to split the batched output of `teacher_predict` to list of output. If `labels_split_fn` is None,
+            `teacher_predict` is expected to return a batched tensor, and it will be auto split on dim 0 by this function.
+            For example, `teacher_predict` returns a tensor with size (32, 100), which is the probability of 100 classes for 32 samples,
+            then after `labels_split_fn`, it will be split to a list of tensor with 32 elements,
+            each tensor is the probability with size (100,).
+        """
+        if self._preprocessed:
+            warn_msg = 'Preprocessor has been generated distillation labels, if more epochs are needed, please set '
+            _logger.warning(warn_msg)
+            return
+
+        labels_split_fn = labels_split_fn if labels_split_fn is not None else self._labels_split_fn
+
+        self._dataset.observe()
+
+        dataloader = self._patched_uid_dataloader()
+        total_batch_num = len(dataloader)
+        log_literal = min(max(total_batch_num // 25, 10), 100)
+
+        for epoch_idx in range(epochs):
+            for batch_idx, packed_batch in enumerate(dataloader):
+                uids, batch = packed_batch
+                soft_labels = self._teacher_predict(batch)
+
+                soft_labels = labels_split_fn(soft_labels)
+
+                assert isinstance(soft_labels, abc.Iterable)
+                list(map(self._storage.record, uids, soft_labels))
+                if log_process and ((batch_idx + 1) % log_literal == 0 or (batch_idx + 1) == total_batch_num):
+                    info_msg = f'[epoch {epoch_idx}: {batch_idx + 1} / {total_batch_num}]'
+                    _logger.info(info_msg)
+
+        # ready for creating replay dataloader
+        self._preprocessed = True
+        self._dataset.replay()
+
+    def create_replay_dataloader(self, labels_collate_fn: Callable[[List], Any] | None = None):
+        """
+        Return a dataloader that concat the original dataset and the generated distillation labels.
+        The iteration of dataloader will return (distil_labels, origin_batch).
+
+        Parameters
+        ----------
+        labels_collate_fn
+            Use to create replayed dataloader.
+            A function used to collate the samples from distillation storage. If `labels_collate_fn` is None,
+            the samples are excepted to be tensors and have same size, they will be batched and return.
+        """
+        labels_collate_fn = labels_collate_fn if labels_collate_fn is not None else _default_labels_collate_fn
```

Comment:
why not use self._labels_collate_fn?

Reply:
nice catch, it should be `self._labels_collate_fn`